modules/apps/portal/portal-mail/src/main/java/com/liferay/portal/mail/sender/internal/messaging/MailSenderMessagingConfigurator.java

```diff
@@ -0,0 +1,109 @@
+ */
+
+package com.liferay.portal.mail.sender.internal.messaging;
+
+import com.liferay.portal.configuration.metatype.bnd.util.ConfigurableUtil;
+import com.liferay.portal.kernel.messaging.Destination;
+import com.liferay.portal.kernel.messaging.DestinationConfiguration;
+import com.liferay.portal.kernel.messaging.DestinationFactory;
+import com.liferay.portal.kernel.util.HashMapDictionaryBuilder;
+import com.liferay.portal.mail.sender.internal.configuration.MailSenderMessagingConfiguration;
+import com.liferay.portal.mail.sender.internal.constants.MailSenderMessagesDestinationNames;
+
+import java.util.Dictionary;
+import java.util.Map;
+
+import org.osgi.framework.BundleContext;
+import org.osgi.framework.ServiceRegistration;
+import org.osgi.service.component.annotations.Activate;
+import org.osgi.service.component.annotations.Component;
+import org.osgi.service.component.annotations.Deactivate;
+import org.osgi.service.component.annotations.Modified;
+import org.osgi.service.component.annotations.Reference;
+
+/**
+ * @author Julius Lee
+ */
+@Component(
+	configurationPid = "com.liferay.portal.mail.sender.internal.configuration.MailSenderMessagingConfiguration",
+	immediate = true, service = {}
+)
+public class MailSenderMessagingConfigurator {
+
+	@Activate
+	protected void activate(
+		BundleContext bundleContext, Map<String, Object> properties) {
+
+		_bundleContext = bundleContext;
+
+		_mailSenderMessagingConfiguration = ConfigurableUtil.createConfigurable(
+			MailSenderMessagingConfiguration.class, properties);
+
+		DestinationConfiguration destinationConfiguration =
+			new DestinationConfiguration(
+				DestinationConfiguration.DESTINATION_TYPE_PARALLEL,
+				MailSenderMessagesDestinationNames.
+					MAIL_SENDER_MESSAGES_PROCESSOR);
+
+		destinationConfiguration.setWorkersCoreSize(
+			_mailSenderMessagingConfiguration.workersCoreSize());
+		destinationConfiguration.setWorkersMaxSize(
+			_mailSenderMessagingConfiguration.workersMaxSize());
+
+		Destination destination = _destinationFactory.createDestination(
+			destinationConfiguration);
+
+		Dictionary<String, Object> dictionary =
+			HashMapDictionaryBuilder.<String, Object>put(
+				"destination.name", destination.getName()
+			).build();
+
+		_serviceRegistration = bundleContext.registerService(
+			Destination.class, destination, dictionary);
+	}
+
+	@Deactivate
+	protected void deactivate() {
+		if (_serviceRegistration != null) {
+			Destination destination = _bundleContext.getService(
+				_serviceRegistration.getReference());
+
+			_serviceRegistration.unregister();
+
+			destination.destroy();
+		}
+
+		_bundleContext = null;
+	}
+
+	@Modified
+	protected void modified(
+		BundleContext bundleContext, Map<String, Object> properties) {
+
+		deactivate();
+
+		activate(bundleContext, properties);
+	}
+
+	private volatile BundleContext _bundleContext;
+
+	@Reference
+	private DestinationFactory _destinationFactory;
+
+	private volatile MailSenderMessagingConfiguration
```

Comment:
@julschong This field is only used in `activate` method, please make it a method variable.